django-backend/fecfiler/transactions/schedule_a/tests/test_serializers.py

```diff
@@ -0,0 +1,157 @@
+
+    def test_serializer_validate(self):
+        valid_serializer = ScheduleATransactionSerializerBase(
+            data=self.valid_schedule_a_transaction,
+            context={"request": self.mock_request},
+        )
+        self.assertTrue(valid_serializer.is_valid(raise_exception=True))
+        invalid_transaction = self.valid_schedule_a_transaction.copy()
+        invalid_transaction["form_type"] = "invalidformtype"
+        del invalid_transaction["contributor_first_name"]
+        invalid_serializer = ScheduleATransactionSerializerBase(
+            data=invalid_transaction,
+            context={"request": self.mock_request},
+        )
+        self.assertFalse(invalid_serializer.is_valid())
+        self.assertIsNotNone(invalid_serializer.errors["form_type"])
+        self.assertIsNotNone(invalid_serializer.errors["contributor_first_name"])
+
+    def test_no_transaction_type_identifier(self):
+        missing_type = self.valid_schedule_a_transaction.copy()
+        del missing_type["transaction_type_identifier"]
+        missing_type_serializer = ScheduleATransactionSerializerBase(
+            data=missing_type,
+            context={"request": self.mock_request},
+        )
+        self.assertFalse(missing_type_serializer.is_valid())
+        self.assertIsNotNone(
+            missing_type_serializer.errors["transaction_type_identifier"]
+        )
+
+    def test_parent(self):
+        parent = self.valid_schedule_a_transaction.copy()
+        parent["transaction_type_identifier"] = "EARMARK_RECEIPT"
+        parent["contribution_purpose_descrip"] = "test"
+        child = self.valid_schedule_a_transaction.copy()
+        child["transaction_type_identifier"] = "EARMARK_MEMO"
+        child["contribution_purpose_descrip"] = "test"
+        child["back_reference_sched_name"] = "test"
+        child["back_reference_tran_id_number"] = "test"
+        child["memo_code"] = True
+        parent["children"] = [child]
+        serializer = ScheduleATransactionSerializer(
+            data=parent,
+            context={"request": self.mock_request},
+        )
+        self.assertTrue(serializer.is_valid(raise_exception=True))
+        serializer.create(serializer.to_internal_value(parent))
+        parent_instance = ScheduleATransaction.objects.filter(
+            transaction_type_identifier="EARMARK_RECEIPT"
+        )[0]
+        # below is not possible with ContentType relation
+        # children = ScheduleATransaction.objects.filter(
+        #     parent_transaction=parent_instance
+        # )
+        children = ScheduleATransaction.objects.filter(
+            parent_transaction_object_id=parent_instance.id
+        )
+        self.assertNotEqual(children.count(), 0)
+
+        parent = parent_instance.__dict__.copy()
+        child = children[0].__dict__.copy()
+        parent["contribution_purpose_descrip"] = "updated parent"
+        child["contribution_purpose_descrip"] = "updated child"
```

Comment:
SonarCloud is reporting a code smell here. It'd like to see a constant for "updated child" string where it occurs in this file.